lib/ansible/plugins/inventory/aws_ssm.py

```diff
@@ -0,0 +1,213 @@
+    requirements:
+    - boto3
+    - botocore
+    short_description: AWS SSM inventory source
+    description:
+      - Fetch parameters via path
+      - Add groups, hosts and variables based on path
+    options:
+      plugin:
+        description: token that ensures this is a source file for the 'aws_ssm' plugin.
+        required: True
+        choices: ['aws_ssm']
+      path:
+        description: Path to search for in ssm
+        required: True
+      region:
+        description: Region to retrieve SSM parameter from
+        required: True
+      decrypt:
+        description: Whether to decrypt secrets or not
+        required: False
+        default: True
+'''
+
+EXAMPLES = '''
+plugin: aws_ssm
+path: prod
+region: us-east-1
+aws_profile: default
+'''
+
+from ansible.errors import AnsibleError
+from ansible.module_utils._text import to_native
+from ansible.module_utils.ec2 import boto3_tag_list_to_ansible_dict
+from ansible.plugins.inventory import BaseInventoryPlugin, Constructable, Cacheable
+from ansible.utils.display import Display
+
+try:
+    import boto3
+    import botocore
+except ImportError:
+    raise AnsibleError("The ssm dynamic inventory plugin requires boto3 and "
+                       "botocore")
+
+
+display = Display()
+
+
+class InventoryModule(BaseInventoryPlugin, Constructable, Cacheable):
+
+    NAME = 'aws_ssm'
+
+    def __init__(self):
+        super(InventoryModule, self).__init__()
+
+    def verify_file(self, path):
+        ''' Verify plugin configuration file. '''
+        if super(InventoryModule, self).verify_file(path):
+            if path.endswith(('aws_ssm.yml', 'aws_ssm.yaml')):
+                return True
+        display.debug("aws_ssm inventory filename must end with 'aws_ssm.yml' "
+                      "or 'aws_ssm.yaml'")
+        return False
+
+    def parse(self, inventory, loader, path, cache=True):
+        ''' Retrieve data '''
+        super(InventoryModule, self).parse(inventory, loader, path)
+        cache_key = self._get_cache_prefix(path)
+        config_data = self._read_config_data(path)
+        self.setup(config_data, cache, cache_key)
+
+    def setup(self, config_data, cache, cache_key):
+        ''' Retrieve parameters from config file and fetch data. '''
+        decrypt = True
+
+        path = config_data.get('path')
+        region = config_data.get('region')
+        if isinstance(config_data.get('decrypt'), bool):
+            decrypt = config_data.get('decrypt')
+
+        if not path or not region:
+            raise AnsibleError('path and region should be present')
+
+        client = self.get_client(config_data)
+
+        source_data = None
+        if cache and cache_key in self._cache:
+            try:
+                source_data = self._cache[cache_key]
+            except KeyError:
+                pass
+
+        if not source_data:
+            self.fetch_data(client, path, decrypt)
+
+    def get_client(self, config_data):
```

Comment:
This is something that should be shared with other inventory plugins. Please move the logic in plugins/inventory/aws_ec2 (or aws_rds) into module_utils and import that instead of doing your own client creation here so we have less duplicate code.

Reply:
I have made use of boto3_inventory_conn from ansible.module_utils.ec2